### PR TITLE
Add `build-test-publish-wheel` CI file

### PR DESCRIPTION
**Pull request by sarahyurick**
Re-opening https://github.com/NVIDIA/NeMo-Curator/pull/353.

### CODE REVIEW

File: .github/workflows/build-test-publish-wheel.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Build, test, and publish a PyPi wheel (to testpypi)
+
+on:
+  pull_request:
```
Comment (sarahyurick) [inline .github/workflows/build-test-publish-wheel.yml:18]:
```suggestion
  push:
```
Will change this back to `push` if the tests succeed.

**Comment by ko3n1g** [comment]
Can you also modify `nemo_curator/__init__.py` like this? https://github.com/NVIDIA/NeMo/blob/main/nemo/__init__.py#L16

File: nemo_curator/package_info.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+MINOR = 6
+PATCH = 0
+PRE_RELEASE = "rc0"
+DEV = "dev1"
+
+# Use the following formatting: (major, minor, patch, pre-release)
+VERSION = (MAJOR, MINOR, PATCH, PRE_RELEASE, DEV)
+
+__shortversion__ = ".".join(map(str, VERSION[:3]))
+__version__ = __shortversion__ + "".join(VERSION[3:])
```
[the remark this answers is not in the document]
Comment (ayushdg) [reply nemo_curator/package_info.py:26]: I'm not completely familiar with versioning semantics but based on what I see from [pypa](https://packaging.python.org/en/latest/specifications/version-specifiers/#developmental-releases), it lines up that we would want the `0.6.0rcN.devN` approach.

File: nemo_curator/package_info.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+MAJOR = 0
+MINOR = 6
+PATCH = 0
+PRE_RELEASE = "rc0"
```
Comment (ayushdg) [inline nemo_curator/package_info.py:19]: Again I'm not sure what the convention is, but I prefer leaving this empty `""` and just having `DEV` until code freeze or QA.

File: .github/workflows/build-test-publish-wheel.yml (hunk 1 above)
Comment (ayushdg) [reply .github/workflows/build-test-publish-wheel.yml:18]: Our dev cycle typically consists of making a `0.n.0` branch around code freeze, and publishing the final release wheel later. Does it make sense to modify the process slightly to create `rc` branches instead?

I wonder if updating the action to trigger on something like is another alternative that works here:
```
on:
  release:
    types: [created]
```

File: nemo_curator/package_info.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarahyurick) [reply nemo_curator/package_info.py:26]: From Oliver:
"
According to PEP 440, the version number of a Python package looks like this: `[N!]N(.N)*[{a|b|rc}N][.postN][.devN]`.

Curator has the `dev` segment but not `rc` (like the other NeMo libraries do). I like that you use `dev` and I could see some use cases in the future where this segment would become very handy. I would like to use this for NeMo Aligner and NeMo itself too. However, `rc` expresses pre-releases so we definitely have to have this too.

`devN` increases (probably with each push to `main`) until a new pre-release resets it.
"

**Comment by sarahyurick** [comment]
Hi @ko3n1g and @ayushdg I have added the changes, please let me know what you think.

File: nemo_curator/package_info.py (hunk 3 above)
Comment (ko3n1g) [reply nemo_curator/package_info.py:19]: No, both segments (prerelease and dev) are filled. For the actual prerelease workflow that tags a commit, we‘ll set dev to "". And for the actual release workflow that tags a commit we‘ll set both dev and prelease to ""

File: .github/workflows/build-test-publish-wheel.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Build, test, and publish a PyPi wheel (to testpypi)
+
+on:
+  push:
```
Comment (ko3n1g) [inline .github/workflows/build-test-publish-wheel.yml:18]:
```suggestion
on:
  pull_request:
  push:
```

**Review by ko3n1g — CHANGES_REQUESTED** [review]
test

File: nemo_curator/__init__.py [hunk 5]
```diff
@@ -12,22 +12,54 @@
...
-    NemoDeployClient,
-    OpenAIClient,
-)
-from .utils.distributed_utils import get_client, get_network_interfaces
+    # Dask will automatically convert the list score type
+    # to a string without this option.
+    # See https://github.com/NVIDIA/NeMo-Curator/issues/33
+    # This also happens when reading and writing to files
+    dask.config.set({"dataframe.convert-string": False})
+except ImportError:
```
Comment (ryantwolf) [inline nemo_curator/__init__.py:23]: Why do we need these import guards?

**Review by ryantwolf — COMMENTED** [review]
Want to dig deeper on the import guards before merging.

File: nemo_curator/__init__.py (hunk 5 above)
Comment (sarahyurick) [reply nemo_curator/__init__.py:23]: We were seeing a Dask `ModuleNotFoundError` in the build and test CI, example here: https://github.com/NVIDIA/NeMo-Curator/actions/runs/11812742210/job/32908443159.

But I agree, we shouldn't have the try/except. Just debugged and figured out that it was an issue with our pyproject.toml file dynamically fetching the version from `nemo_curator/__init__.py` - switching it to fetch from the `package_info.py` file should work. Will push changes shortly.

Comment (ryantwolf) [reply nemo_curator/__init__.py]: Thank you!

**Review by ryantwolf — APPROVED** [review]
Looks good now that the import guard is gone.

**Review by ko3n1g — APPROVED** [review]

**Pull request merged**